### PR TITLE
📝Xref line info

### PR DESCRIPTION
**Pull request by live1206**
- Add line info for xref related errors
- Add test for TOC bookmark validation

### CODE REVIEW

File: docs/specs/toc.yml [hunk 1]
```diff
@@ -1106,21 +1106,25 @@ outputs:
...
-  docs/TOC.md: |	
-    # [link to title 1](a.md#title-1)	
-outputs:	
-  docs/a.json:	
-  docs/toc.json: |	
-    {	
-      "items": [	
-        { "name": "link to title 1", "href": "a#title-1", "!monikers":null },	
-      ]	
+# validate bookmarks in toc
```
Comment (OsmondJiang) [inline docs/specs/toc.yml]: would you split this to another PR?

Comment (live1206) [reply docs/specs/toc.yml]: Just added this test case back, the implementation was done in the previous PR

File: src/docfx/build/page/AttributeTransformer.cs [hunk 2]
```diff
@@ -12,44 +12,43 @@ namespace Microsoft.Docs.Build
...
+            object Transform(IEnumerable<DataTypeAttribute> attributes, object value, string jsonPath, Range range)
             {
                 var attribute = attributes.SingleOrDefault(attr => !(attr is XrefPropertyAttribute));
-                return TransformContent(context, attribute, value, file, buildChild);
+                return TransformContent(context, attribute, value, range, file, buildChild);
             }
         }
 
-        public static Func<IEnumerable<DataTypeAttribute>, object, string, object> TransformXref(
+        public static Func<IEnumerable<DataTypeAttribute>, object, string, Range, object> TransformXref(
```
Comment (OsmondJiang) [inline src/docfx/build/page/AttributeTransformer.cs]: Suggest to:
```
 Func<IEnumerable<DataTypeAttribute>, SourceInfo<object>, string, object>
```

Comment (OsmondJiang) [reply src/docfx/build/page/AttributeTransformer.cs]: and finally we can get ride of range, haha

File: src/docfx/lib/log/SourceInfo.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ public sealed class SourceInfo<T> : ISourceInfo
 
         object ISourceInfo.Value => Value;
 
-        public SourceInfo(T value, string file, Range range)
+        public SourceInfo(T value, Range range = default, string file = null)
```
Comment (superyyrrzz) [inline src/docfx/lib/log/SourceInfo.cs]: As `SourceInfo` has `file` already, seems we don't need `Document file` as a parameter in most methods of `errors.cs`.

Comment (live1206) [reply src/docfx/lib/log/SourceInfo.cs]: will clean this up after we merge #4311
